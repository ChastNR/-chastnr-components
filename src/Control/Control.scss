.ctrl {
  margin-bottom: 16px;

  .ctrl__cntr {
    background-color: #ecf1f7;
    border-radius: 8px;
    border: 1px solid #ecf1f7;
    box-sizing: border-box;
    color: #333;
    cursor: text;
    display: flex;
    flex-direction: column;
    font-size: 15px;
    font-weight: 400;
    height: 56px;
    letter-spacing: 0;
    line-height: 16px;
    max-width: 450px;
    outline: 0;
    padding: 15px;
    position: relative;
    transition-duration: 0.25s;
    transition-property: background-color, border-color, margin-bottom;
    transition-timing-function: ease;
    width: 100%;
    word-wrap: break-word;

    .ctrl__lb {
      color: rgba(0, 0, 0, 0.54);
      cursor: inherit;
      display: block;
      display: inline-block;
      font-family: inherit;
      font-weight: 400;
      line-height: 1.8;
      margin-bottom: 0.5rem;
      margin: 0;

      @media (max-width: 575.98px) {
        line-height: 1.5;
      }
    }

    > * {
      font-family: inherit;
    }

    .ctrl__cnt {
      align-items: flex-end;
      align-self: stretch;
      display: flex;
      flex-direction: column;
      justify-content: center;
      margin-right: 15px;
      position: absolute;
      right: 0;
    }

    .ctrl__lb {
      font-size: 15px;
      font-weight: 400;
      letter-spacing: 0;
      line-height: 24px;
      padding: 15px 0;
      position: absolute;
      top: 0;
      transform: scale(1) translateY(0);
      transition: all 0.25s ease;
      user-select: none;
      word-wrap: break-word;
      z-index: 1;
    }

    &:hover {
      border-color: #e4ebf3;
      background-color: #e4ebf3;

      &.error {
        border-color: transparent;
        background-color: rgba(224, 31, 25, 0.16);
      }
    }

    &:focus-within {
      border-color: #616871;
      background: #fff;

      .ctrl__lb {
        color: rgba(0, 0, 0, 0.8);
        transform: translate(-6.66667%, -10.38462px) scale(0.86667);
      }

      &.error {
        background: #fff;
        border-color: #f52222;

        &.hasValue {
          .ctrl__lb {
            color: #e31c1c;
          }
        }
      }
    }

    &.disabled {
      opacity: 0.56;
      pointer-events: none;
    }

    &.hasValue {
      .ctrl__lb {
        color: rgba(0, 0, 0, 0.8);
        transform: translate(-6.66667%, -10.38462px) scale(0.86667);
      }
    }

    &.error {
      background-color: rgba(224, 31, 25, 0.12);
      border: 1px solid transparent;

      &.hasValue {
        .ctrl__lb {
          color: #e31c1c;
        }
      }
    }

    &.required {
      .ctrl__lb {
        &:after {
          content: '*';
        }
      }
    }
  }
}
