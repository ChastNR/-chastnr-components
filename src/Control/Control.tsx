import { forwardRef, PropsWithChildren } from "react";

import classNames from "clsx";

import Error from "../Error";

import "./Control.scss";

interface ControlProps {
  className?: string;
  content?: React.ReactNode;
  disabled?: boolean;
  name?: string;
  error?: React.ReactNode;
  hasValue: boolean;
  label: React.ReactNode;
  onBlur?: React.FocusEventHandler<HTMLDivElement>;
  onFocus?: React.FocusEventHandler<HTMLDivElement>;
  required?: boolean;
  role?: React.AriaRole;
}

const Control = forwardRef<HTMLDivElement, PropsWithChildren<ControlProps>>(
  (
    {
      children,
      className,
      content,
      disabled,
      error,
      hasValue,
      label,
      name,
      onBlur,
      onFocus,
      required,
    },
    ref
  ) => {
    const controlClassNames = classNames("ctrl__cntr", className, {
      required,
      disabled,
      error,
      hasValue,
    });

    return (
      <div className="ctrl">
        <div
          className={controlClassNames}
          onBlur={onBlur}
          onFocus={onFocus}
          ref={ref}
          role="textbox"
          tabIndex={0}
        >
          <label className="ctrl__lb" htmlFor={name}>
            {label}
          </label>
          {children}
          {content && <div className="ctrl__cnt">{content}</div>}
        </div>
        <Error>{error}</Error>
      </div>
    );
  }
);

export default Control;
