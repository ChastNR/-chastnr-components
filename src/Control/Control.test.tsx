import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";

import Control from "./Control";

describe("<Control />", () => {
  it("should render", () => {
    const control = shallow(<Control hasValue={false} label="testLabel" />);
    expect(toJson(control)).toMatchSnapshot();
  });

  it("props should be equal", () => {
    const className = "testClassName";
    const content = "testContent";
    const name = "testName";
    const disabled = true;
    const error = "testError";
    const label = "testLabel";
    const required = true;
    const hasValue = true;
    const role = "textbox";

    const control = mount(
      <Control
        className={className}
        content={content}
        disabled={disabled}
        error={error}
        hasValue={hasValue}
        label={label}
        name={name}
        required={required}
        role={role}
      />
    );

    expect(control.props().className).toEqual(className);
    expect(control.props().content).toEqual(content);
    expect(control.props().name).toEqual(name);
    expect(control.props().disabled).toEqual(disabled);
    expect(control.props().error).toEqual(error);
    expect(control.props().label).toEqual(label);
    expect(control.props().required).toEqual(required);
    expect(control.props().hasValue).toEqual(hasValue);
    expect(control.props().role).toEqual(role);
  });
});
