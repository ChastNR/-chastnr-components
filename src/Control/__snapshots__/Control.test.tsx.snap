// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Control /> should render 1`] = `
<div
  className="ctrl"
>
  <div
    className="ctrl__cntr"
    role="textbox"
    tabIndex={0}
  >
    <label
      className="ctrl__lb"
    >
      testLabel
    </label>
  </div>
  <Error />
</div>
`;
